libs/Microsoft.MixedReality.WebRTC/ISignalingService.cs

```diff
@@ -0,0 +1,221 @@
+        {
+            throw new NotImplementedException();
+        }
+
+        public void Start(string localEndPointId)
+        {
+            throw new NotImplementedException();
+        }
+
+        public void Stop()
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    static class SignalingExamples
+    {
+        // Use the service to connect a local PeerConnection to another chosen by the user among the available ones.
+        public static void ChooseAndConnect(PeerConnection pc)
+        {
+            using (ISignalingService service = new PeerSignalingService(IPAddress.Broadcast, 1234, "MyApp"))
+            {
+                service.Start("LocalEndpoint#123");
+
+                string target = null;
+                while (target == null)
+                {
+                    target = MakeTheUserChoose(service.RemoteEndPoints);
+                }
+
+                string sessionId = Guid.NewGuid().ToString();
+
+                // We have a target
+                pc.LocalSdpReadytoSend +=
+                    (string type, string sdp) =>
+                    {
+                        service.SendToRemoteEndpoint(target,
+                            new SignalingMessage
+                            {
+                                SessionId = sessionId,
+                                Type = FromSMTString(type),
+                                Payload = sdp
+                            });
+                    };
+                pc.IceCandidateReadytoSend +=
+                    (string candidate, int sdpMlineindex, string sdpMid) =>
+                    {
+                        string payload = candidate + '|' + sdpMlineindex + '|' + sdpMid;
+                        service.SendToRemoteEndpoint(target,
+                            new SignalingMessage
+                            {
+                                SessionId = sessionId,
+                                Type = SignalingType.IceCandidate,
+                                Payload = payload
+                            });
+                    };
+                service.MessageReceived +=
+                    (senderId, message) =>
+                    {
+                        if (senderId == target && message.SessionId == sessionId)
+                        {
+                            if (message.Type == SignalingType.SdpAnswer)
+                            {
+                                pc.SetRemoteDescription("answer", message.Payload);
+                            }
+                            else if (message.Type == SignalingType.IceCandidate)
+                            {
+                                var parts = message.Payload.Split('|');
+                                pc.AddIceCandidate(parts[2], int.Parse(parts[1]), parts[0]);
+                            }
+                        }
+                    };
+                pc.CreateOffer();
```

Comment:
I think either `CreateOffer()` should be removed, or you should go to the full length and see the connection through by also waiting for the answer and the connected event. This alone starts connecting but terminates the method while the connection is pending.

Reply:
Yes, I omitted it but that's necessary. I assume this should also listen to `RenegotiationNeeded` to create new offers.

Reply:
I have taken the responsibility of creating offers out of the utility, now the utility only forwards outgoing/incoming messages, and the responsibility of initiating/reinitiating the negotiation is on the caller. It seems a bit cleaner this way, but I am willing to discuss this.